### DISCUSSION TITLE
making-getoradd-on-concurrentdictionary-thread-safe-using-lazy/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#55] [Announcements]
# making-getoradd-on-concurrentdictionary-thread-safe-using-lazy/

This post explores the GetOrAdd function of ConcurrentDictionary, the level of thread safety it provides, and ways to add additional threading constraints.

https://andrewlock.net/making-getoradd-on-concurrentdictionary-thread-safe-using-lazy/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at November 17 2016, 07:15, in reply to [alastairtree](https://disqus.com/by/alastairtree/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/55#discussioncomment-6024606)</em>

---

<p>Looks like a very handy little library, thanks for sharing:)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 19 2017, 08:11, in reply to [troplin](https://disqus.com/by/troplin/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/55#discussioncomment-6024611)</em>

---

<p>Yes, Lazy initialisation can throw, and depending on how you have configured it, you can either cache that exception, or have subsequent attempts retry the create until one succeeds. See here for details: <a href="https://msdn.microsoft.com/en-us/library/dd997286(v=vs.110).aspx#Anchor_1" rel="nofollow noopener" title="https://msdn.microsoft.com/en-us/library/dd997286(v=vs.110).aspx#Anchor_1">https://msdn.microsoft.com/en-us/library/dd997286(v=vs.110).aspx#Anchor_1</a></p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at July 27 2019, 07:20, in reply to [Andrew Harward](https://disqus.com/by/andrewharward/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/55#discussioncomment-6024613)</em>

---

<p>As described in the link above, you can not cache exceptions if you want by changing constructor arguments as described in the docs: <a href="https://docs.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization#exceptions-in-lazy-objects" rel="nofollow noopener" title="https://docs.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization#exceptions-in-lazy-objects">https://docs.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization#exceptions-in-lazy-objects</a></p>

[a comment by an automated account is not shown]

**Comment by MoonStorm** [comment]
Ouch, what a bad suggestion. You've just moved the problem somewhere else. 

Lazy by default uses ExecutionAndPublication mode, which means the value factory gets run once, but it "caches" the result internally, whether good or bad. That changes the way the value factory gets otherwise normally used by the ConcurrentDictionary in case the factory fails. For what they've done in that MVC pipeline, this is perfectly fine. In most cases though, that's undesirable. 

To avoid this behavior you might be tempted to switch the Lazy mode to PublicationOnly, which pretty much puts you back at square one, where the value factory can run multiple times. 

When it comes to factories returning IDisposable, people should NOT use the GetOrAdd altogether, but a mixture of TryGetValue and, in case of a negative outcome, by an explicit MANUAL construction of the value, followed by TryAdd and a Value.Dispose in case someone else beat you to the punch by testing the returned value with a ReferenceEquals against your prepared new value.

As an alternative, write your code with a regular Dictionary with guarded access via ReaderWriterLockSlim in case of high number of writes OR via a simple lock in case of low number of writes. That's it.

**Reply by MoonStorm** [reply]
In other words:

```
            MyExpensiveResource finalExpensiveResource = null;

            if (!concurrentCollection.TryGetValue(key, out finalExpensiveResource))
            {
                // try again, but under the assumption that the collection doesn't have it
                var newExpensiveResource = new MyExpensiveResource();
                finalExpensiveResource = concurrentCollection.GetOrAdd(key, newExpensiveResource);
                
                if (!ReferenceEquals(newExpensiveResource, finalExpensiveResource))
                {
                    // our new resource was NOT used, somebody else beat us to the punch
                    newExpensiveResource.Dispose();
                }
            }

```
